src/main/java/jcifs/dcerpc/msrpc/lsarpc.java

```diff
@@ -0,0 +1,1046 @@
+                this.root_directory.encode(_dst);
+
+            }
+            if ( this.object_name != null ) {
+                _dst = _dst.deferred;
+                this.object_name.encode(_dst);
+
+            }
+            if ( this.security_quality_of_service != null ) {
+                _dst = _dst.deferred;
+                this.security_quality_of_service.encode(_dst);
+
+            }
+        }
+
+
+        @Override
+        public void decode ( NdrBuffer _src ) throws NdrException {
+            _src.align(4);
+            this.length = _src.dec_ndr_long();
+            int _root_directoryp = _src.dec_ndr_long();
+            int _object_namep = _src.dec_ndr_long();
+            this.attributes = _src.dec_ndr_long();
+            this.security_descriptor = _src.dec_ndr_long();
+            int _security_quality_of_servicep = _src.dec_ndr_long();
+
+            if ( _root_directoryp != 0 ) {
+                _src = _src.deferred;
+                this.root_directory.decode(_src);
+
+            }
+            if ( _object_namep != 0 ) {
+                if ( this.object_name == null ) { /* YOYOYO */
+                    this.object_name = new rpc.unicode_string();
+                }
+                _src = _src.deferred;
+                this.object_name.decode(_src);
+
+            }
+            if ( _security_quality_of_servicep != 0 ) {
+                if ( this.security_quality_of_service == null ) { /* YOYOYO */
+                    this.security_quality_of_service = new LsarQosInfo();
+                }
+                _src = _src.deferred;
+                this.security_quality_of_service.decode(_src);
+
+            }
+        }
+    }
+
+    public static class LsarDomainInfo extends NdrObject {
+
+        public rpc.unicode_string name;
+        public rpc.sid_t sid;
+
+
+        @Override
+        public void encode ( NdrBuffer _dst ) throws NdrException {
+            _dst.align(4);
+            _dst.enc_ndr_short(this.name.length);
+            _dst.enc_ndr_short(this.name.maximum_length);
+            _dst.enc_ndr_referent(this.name.buffer, 1);
+            _dst.enc_ndr_referent(this.sid, 1);
+
+            if ( this.name.buffer != null ) {
+                _dst = _dst.deferred;
+                int _name_bufferl = this.name.length / 2;
+                int _name_buffers = this.name.maximum_length / 2;
+                _dst.enc_ndr_long(_name_buffers);
+                _dst.enc_ndr_long(0);
+                _dst.enc_ndr_long(_name_bufferl);
+                int _name_bufferi = _dst.index;
+                _dst.advance(2 * _name_bufferl);
+
+                _dst = _dst.derive(_name_bufferi);
+                for ( int _i = 0; _i < _name_bufferl; _i++ ) {
+                    _dst.enc_ndr_short(this.name.buffer[ _i ]);
+                }
+            }
+            if ( this.sid != null ) {
+                _dst = _dst.deferred;
+                this.sid.encode(_dst);
+
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 129:154, 595:620.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.